smart-contracts/contracts/mixins/MixinPurchase.sol

```diff
@@ -84,28 +84,56 @@ contract MixinPurchase is
     uint inMemoryKeyPrice = keyPrice;
     (discount, tokens) = unlockProtocol.computeAvailableDiscountFor(_recipient, inMemoryKeyPrice);
 
+    if (discount > 0) {
+      // Note that this does not include any discount provided by the hook
+      unlockProtocol.recordConsumedDiscount(discount, tokens);
+    }
+
+    discount += _onKeySold(_recipient, _referrer, inMemoryKeyPrice, _data);
+
     if (discount > inMemoryKeyPrice) {
       inMemoryKeyPrice = 0;
     } else {
       // SafeSub not required as the if statement already confirmed `inMemoryKeyPrice - discount` cannot underflow
       inMemoryKeyPrice -= discount;
     }
 
-    if (discount > 0) {
-      unlockProtocol.recordConsumedDiscount(discount, tokens);
-    }
-
-    unlockProtocol.recordKeyPurchase(inMemoryKeyPrice, getHasValidKey(_referrer) ? _referrer : address(0));
-
     // We explicitly allow for greater amounts of ETH or tokens to allow 'donations'
     if(tokenAddress != address(0)) {
       require(_value >= inMemoryKeyPrice, 'INSUFFICIENT_VALUE');
       inMemoryKeyPrice = _value;
     }
-    // Security: after state changes to minimize risk of re-entrancy
```

Comment:
I'm not so sure what the re-entrancy risks are at this point. I can't think of an attack but maybe we should add a `ReentrancyGuard` to most calls just in case.